ggshield/oauth.py

```diff
@@ -0,0 +1,273 @@
+import os
+import urllib.parse as urlparse
+import webbrowser
+from base64 import urlsafe_b64encode
+from datetime import datetime
+from hashlib import sha256
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from secrets import token_urlsafe
+from typing import Any, Dict, Optional, Type
+
+import click
+import requests
+from oauthlib.oauth2 import OAuth2Error, WebApplicationClient
+
+from .client import retrieve_client
+from .config import AccountConfig, Config
+
+
+CLIENT_ID = "ggshield_oauth"
+SCOPE = "scan"
+REDIRECT_PORT = 1234
+
+
+class OAuthClient:
+    def __init__(self, config: Config, instance: str) -> None:
+        self.config = config
+        self.instance = instance
+        self._oauth_client = WebApplicationClient(CLIENT_ID)
+        self._access_token: Optional[str] = None
+
+        self.server: Optional[HTTPServer] = None
+        self._handler_wrapper = RequestHandlerWrapper(oauth_client=self)
+        self._generate_pkce_pair()
+
+    def oauth_process(self) -> None:
+        """
+        Handle the whole oauth process which includes
+        - opening the user's webbrowser to GitGuardian login page
+        - open a server and wait for the callback processing
+        """
+        # enable redirection to http://localhost
+        os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = str(True)
+
+        self._token_name = "ggshield token " + datetime.today().strftime("%Y-%m-%d")
+
+        self._prepare_server()
+        self._redirect_to_login()
+        self._wait_for_code()
+        click.echo(f"Created Personal Access Token {self._token_name}")
```

Comment:
It does not feel very explicit what this function does, both reading the docstring and the code.

I don't expect `_wait_for_code` to do much but it appears to even create the token. There are some names to improve.